### ISSUE TITLE
Marker edges are randomly missing

### ISSUE THREAD

**Issue by cpitclaudel** [matplotlib/matplotlib#4083]
The following code, run with version 1.3 or with 1.5.dev1, sometimes produces an entirely blank plot. This happens randomly; rerunning the script multiple times eventually produces the right output. This is on Linux, running python 3.4.0.

If the markers have a color, that color always appears. The edge sometimes are there, and sometimes aren't.

```python
from matplotlib import pyplot

xs = [x for x in range(10)] * 10
ys = [x % 27 for x in range(10*10)]

pyplot.scatter(xs, ys, s=100, marker='o', color="none", edgecolor="red")
pyplot.savefig("test.png")
```

Is there a workaround? Can I help diagnose this bug? The two pictures below were produced by running the exact same program.

![dots](https://cloud.githubusercontent.com/assets/8181630/6096982/f4f95404-af78-11e4-850f-c609f7c3efca.png)
![no-dots](https://cloud.githubusercontent.com/assets/8181630/6096981/f4f767b6-af78-11e4-8cb5-d431e5f28a38.png)

**Comment by cpitclaudel** [comment]
Here are two traces (`strace python3 test.py`): [edges appearing properly](http://pit-claudel.fr/clement/matplotlib/dots.trace) | [edges missing](http://pit-claudel.fr/clement/matplotlib/no-dots.trace)

And the two pdfs: [edges appearing properly](http://pit-claudel.fr/clement/matplotlib/dots.pdf) | [edges missing](http://pit-claudel.fr/clement/matplotlib/no-dots.pdf)

**Comment by cpitclaudel** [comment]
Extra note: This seems to happen independently of the backend

**Comment by cpitclaudel** [comment]
And this doesn't seem to happen when using python 2.7

**Comment by tacaswell** [comment]
My guess is that you are running into issues with the state-machine and not drawing to the figure you think you are drawing to.

Can you reporduce this with 

```python
from matplotlib import pyplot as plt

xs = [x for x in range(10)] * 10
ys = [x % 27 for x in range(10*10)]
fig, ax = plt.subplots()
ax.scatter(xs, ys, s=100, marker='o', color="none", edgecolor="red")
fig.savefig("test.png")
```

**Comment by cpitclaudel** [comment]
Thanks for your answer. I get exactly the same results with your code: sometimes I get circles; sometimes not.

**Comment by efiring** [comment]
@cpitclaudel What is your rough estimate of the frequency: is one outcome much more common, or do they seem equally likely?  (It sounds like a timing issue, as if the plotting were being done in a thread and sometimes the file is written before the thread finishes--but of course this isn't it, because we aren't doing anything like that.  Everything should be strictly sequential.)

**Comment by WeatherGod** [comment]
The observation that this is only happening with py3.4 makes me very
suspicious. Of what? I haven't a clue, but something is fishy. We aren't
seeing these sorts of issues with Travis testing.  Could we simplify the
code any further? Or is this strictly a problem with marker edges?

**Comment by tacaswell** [comment]
I can not reproduce this at all with py34

```python

import matplotlib
import numpy as np
from PIL import Image

matplotlib.use('agg')
from matplotlib import pyplot as plt

xs = [x for x in range(10)] * 10
ys = [x % 27 for x in range(10*10)]

N = 500

for j in range(N):
    fig, ax = plt.subplots()
    ax.scatter(xs, ys, s=100, marker='o', color="none", edgecolor="red")
    fig.savefig("test_{:05d}.png".format(j))
    plt.close('all')

base_image = np.array(Image.open('test_00000.png'))
for j in range(N):
    test_image = np.array(Image.open("test_{:05d}.png".format(j)))
    if np.max(base_image - test_image):
        print("frame {} is bad".format(j))

```

does not print anything (which is confirmed by visually scanning the thumbnails). 

What else is going on in your scripts?

**Comment by cpitclaudel** [comment]
I think about equally likely. I can reproduce this on two distinct Linux Mint machines

**Comment by cpitclaudel** [comment]
Could this be due to a memory corruption somewhere instead of a timing issue? Valgrind produces too many errors to be useful.

**Comment by cpitclaudel** [comment]
There's nothing else going on in my scripts that the MWE I posted. I'm not sure if I can reduce it further The problem also appears with 3D scatter plots.

**Comment by cpitclaudel** [comment]
@tacaswell I ran your script. It never returns any output; the images are always consistent. Sometimes it makes 500 blank images, sometimes 500 images with circles.

**Comment by tacaswell** [comment]
I just re-ran my script and got 5k frames with _no_ points shown....

The dependency on the process makes me think than _someplace_ we are relying on ordering in a dictionary.  Making key ordering in a dictionary random is one of the python 3 related changes iirc (to avoid DoS attacks on dictionaries).

**Comment by cpitclaudel** [comment]
Ah, a repro! I'm glad (and a bit sorry; one might have hoped my machines were just crazy). The dictionary is something I thought about earlier too, but I wasn't sure how to force the order of keyword arguments (`ordereddict`s didn't seem to help) :/ I didn't realize that python2 didn't behave the same way.

**Comment by cpitclaudel** [comment]
Some extra info: the machine I'm testing this on has a lot of ram (32GB) and a very recent processor.

**Comment by tacaswell** [comment]
Ah! Great it is not as bad as I had feared

```python

for j in range(N):
    fig, ax = plt.subplots()
    sc = ax.scatter(xs, ys, s=100, marker='o', color="none", edgecolor="red")
    print('ec: ', sc.get_edgecolor())
    fig.savefig("test_{:05d}.png".format(j))
    plt.close('all')
```

sometimes reports no edgcolor.  This means only that `ax.scatter` is broken, not our whole draw stack!

**Comment by efiring** [comment]
@tacaswell, Is it worth seeing if you get the same result with #4079?

**Comment by tacaswell** [comment]
Yes, I was just thinking about that.

**Comment by efiring** [comment]
I left in something I still don't like--the blanket update with kwargs, some of which have already been applied.  What might help, though, is that I put edgecolors in explicitly.

Here's another clue: I put in "edgecolors" only; but in the example, "edgecolor" is going in, which is accepted as an alias in the PathCollection.  If nothing else, I have to do something about this in #4079. We need some kwarg pre-processing here.

**Comment by tacaswell** [comment]
I am inclined to say that the fix for this needs to be stuck on top of the 1.4.3 rc.

**Comment by tacaswell** [comment]
Yup, the dict key ordering is the issue and 

```python
sc = ax.scatter(xs, ys, s=100, marker='o', edgecolors="red", c="none")
```

seems to consistently work.

**Comment by cpitclaudel** [comment]
@tacaswell Indeed! It doesn't work with `color` instead of  `c`, though.

**Comment by tacaswell** [comment]
```python
sc = ax.scatter(xs, ys, s=100, marker='o', edgecolors="red", facecolor="none")
```

should also work consistently.

The issue is that the kwargs `color`  and `edgecolor` are not explicitly handled by the `ax.scatter` method (but the kwarg `c` _is_) and are instead handled by a call to `collection.update(kwargs)` which in turn calls `coll.set_color` and `coll.set_edgecolor`.  `set_color` sets both the face and edge color to be the same (in this case 'none').  If the keys came out in the order `['color', 'edgecolor']` it worked fine as the edge color overrides the empty color and your graph looks fine.  If they come out in the other order the `color` value overrides the edge color and you get invisible dots.

I take back what I said about back-porting this to 1.4.3.  It as a really annoying degeneracy in the input, but not a show stopper.

**Comment by cpitclaudel** [comment]
Indeed, that's correct! Fascinating bug. Thanks for the explanation.

**Comment by tacaswell** [comment]
This is fixed by #4079
